src/audio/module_adapter/module_adapter.c

```diff
@@ -463,6 +463,47 @@ static void module_copy_samples(struct comp_dev *dev, struct comp_buffer *src_bu
 	comp_update_buffer_consume(src_buffer, copy_bytes);
 }
 
+static void module_adapter_process_output(struct comp_dev *dev)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct comp_buffer *sink;
+	struct list_item *blist;
+	int i;
+
+	/* copy all produced output samples to output buffers */
+	i = 0;
+	list_for_item(blist, &mod->sink_buffer_list) {
+		struct comp_buffer *buffer = container_of(blist, struct comp_buffer, sink_list);
+
+		if (mod->output_buffers[i].size > 0) {
+			ca_copy_from_module_to_sink(&buffer->stream, mod->output_buffers[i].data,
+						    mod->output_buffers[i].size);
+			audio_stream_produce(&buffer->stream, mod->output_buffers[i].size);
+		}
+		i++;
+	}
+
+	/* copy from all output local buffers to sink buffers */
+	i = 0;
+	list_for_item(blist, &dev->bsink_list) {
+		struct comp_buffer *src_buffer;
+		struct list_item *_blist;
+		int j = 0;
+
+		sink = container_of(blist, struct comp_buffer, source_list);
+		list_for_item(_blist, &mod->sink_buffer_list) {
```

Comment:
in practice this probably won't run too many iterations, maybe just one or two, but since you're optimising here... You could also try to optimise this loop out. You only need a single `i == j` entry. In fact you just need an `src_buffer` corresponding to `mod->output_buffers[i]`. Maybe it would be possible to just add a pointer to that buffer to `struct output_stream_buffer`?

Reply:
@lyakh this is a good idea indeed. But the struct output_stream_buffer has been defined to maintain compatibility with windows. @mwasko do you think it would be OK to add the extra field into this struct? I will leave this as a future optimization once I get confirmation.

Reply:
@jxstelter can you take a look at that and comment?

Reply:
Adding additional field in at the end of struct output_stream_buffer will probably not break the compatibility, however the question is when this filed src_buffer will be updated? The 3rd party modules will not touch this field at all.  In our draft implementation we have in module_adapter/module/ shim layer for windows modules defined. There are some small windows libraries specific actions implemented. Maybe this could be the place for setting this src_buffer pointer?

Reply:
@jxstelter this field will be set during prepare() when all the intermediate buffers are allocated. In fact, for most components (that produce period_bytes every period), this wont even be used.